controllers/cloudstackfailuredomain_controller.go

```diff
@@ -130,25 +152,229 @@ func (r *CloudStackFailureDomainReconciliationRunner) ClearMachines() (ctrl.Resu
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+
+				// patch etcdadmcluster to refer newly created cloudstack machine template, which will trigger an etcdadmcluster rolling upgrade.
+				if _, err := controllerutil.CreateOrPatch(r.RequestCtx, r.K8sClient, etcdadmCluster,
+					func() error {
+						return unstructured.SetNestedField(
+							etcdadmCluster.Object,
+							newCsMachineTemplateName,
+							"spec", "infrastructureTemplate", "name",
+						)
+					}); err != nil {
 					return ctrl.Result{}, err
 				}
+				r.Log.Info(fmt.Sprintf("failuredomain delete: etcdadmcluster infrastructureTemplate cloudstackmachinetemplate name %s patched", newCsMachineTemplateName))
+				break
 			}
 		}
+		if etcdMachineFound {
+			break
+		}
 	}
-	if len(machines.Items) > 0 {
-		return r.RequeueWithMessage("FailureDomain still has machine(s) in it.")
+
+	return ctrl.Result{}, nil
+}
+func triggerMachineDeploymentRollout(machines []infrav1.CloudStackMachine, r *CloudStackFailureDomainReconciliationRunner) (ctrl.Result, error) {
+	for _, machine := range machines {
+		for _, ref := range machine.OwnerReferences {
+			if ref.Kind == kindMachineSet {
+
+				// get machine deployment by using cloudstack machine's label info
+				md := &clusterv1.MachineDeployment{}
+				mdName, ok := machine.Labels[clusterv1.MachineDeploymentLabelName]
+				if !ok {
+					return ctrl.Result{}, errors.Errorf("cloudstack machine %s misses label: %s", machine.Name, clusterv1.MachineDeploymentLabelName)
+				}
+
+				if err := r.K8sClient.Get(r.RequestCtx, client.ObjectKey{Namespace: machine.Namespace, Name: mdName}, md); err != nil {
+					return ctrl.Result{}, err
+				}
+
+				// add an annotation restartedAt in machine deployment if such one not already added
+				// this will trigger an immediate machine deployment rollout
+				patcher, err := patch.NewHelper(md, r.K8sClient)
+				if md.Spec.Template.Annotations == nil {
+					md.Spec.Template.Annotations = map[string]string{}
+				}
+				timeNowStr := time.Now().Add(time.Second * time.Duration(10)).Format(time.RFC3339)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+				md.Spec.Template.Annotations["cluster.x-k8s.io/restartedAt"] = timeNowStr
+				if err := patcher.Patch(r.RequestCtx, md); err != nil {
+					return ctrl.Result{}, err
+				}
+				r.Log.Info(fmt.Sprintf("failuredomain delete: machine deployment name %s add annotation restartedAt %s", mdName, timeNowStr))
+				break
+			}
+		}
 	}
 	return ctrl.Result{}, nil
 }
 
+func triggerControlPlaneRollout(machines []infrav1.CloudStackMachine, r *CloudStackFailureDomainReconciliationRunner) (ctrl.Result, error) {
+	var cpMachine infrav1.CloudStackMachine
+	var ownerRef metav1.OwnerReference
+	cpMachineFound := false
+	for _, machine := range machines {
+		for _, ref := range machine.OwnerReferences {
+			if ref.Kind == kindKubeadmControlPlane {
+				cpMachineFound = true
+				cpMachine = machine
+				ownerRef = ref
+			}
```

Comment:
It's surprising to me that this doesn't break on the first found CP machine.  Reading further, I see it wants to check whether there are any etcd machines, so it makes sense that it presses on.  I guess this means, then, that it acts on the *last* control plane machine found.  Which I imagine is fine.  But I'd suggest comments to make it clear that that's your intent, since that's a less typical approach.